File: frontend/src/app/core/services/backlog.service.ts
import { Injectable } from '@angular/core';
import { Observable, of, throwError } from 'rxjs';
import { BacklogItem } from '../models/backlog-item.model';
import { BacklogCategory } from '../enums/enums';

/** localStorage service for all backlog-item operations. */
@Injectable({ providedIn: 'root' })
export class BacklogService {
    private readonly STORAGE_KEY = 'wpt_backlog_items';
    private readonly ASSIGNMENTS_KEY = 'wpt_assignments';
    private readonly PLAN_KEY = 'wpt_current_plan';

    private readAll(): BacklogItem[] {
        const raw = localStorage.getItem(this.STORAGE_KEY);
        return raw ? JSON.parse(raw) as BacklogItem[] : [];
    }

    private writeAll(items: BacklogItem[]): void {
        localStorage.setItem(this.STORAGE_KEY, JSON.stringify(items));
    }

    private uuid(): string {
        return crypto.randomUUID ? crypto.randomUUID() : 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
            const r = Math.random() * 16 | 0;
            return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
        });
    }

    /** Get the set of backlog item IDs that are assigned in the current active plan. */
    private getInPlanIds(): Set<string> {
        const plan = localStorage.getItem(this.PLAN_KEY);
        if (!plan) return new Set();
        const currentPlan = JSON.parse(plan);
        const raw = localStorage.getItem(this.ASSIGNMENTS_KEY);
        const assignments: any[] = raw ? JSON.parse(raw) : [];
        const weekAssignments = assignments.filter(a => a.weeklyPlanId === currentPlan.id);
        return new Set(weekAssignments.map(a => a.backlogItemId));
    }

    getAll(includeArchived = false, category?: BacklogCategory, search?: string): Observable<BacklogItem[]> {
        let items = this.readAll();
        const inPlanIds = this.getInPlanIds();

        // Apply filters
        if (!includeArchived) {
            items = items.filter(i => !i.isArchived);
        }
        if (category) {
            items = items.filter(i => i.category === category);
        }
        if (search && search.trim()) {
            const s = search.trim().toLowerCase();
            items = items.filter(i => i.title.toLowerCase().includes(s));
        }

        // Set isInActivePlan flag
        items = items.map(i => ({ ...i, isInActivePlan: inPlanIds.has(i.id) }));

        return of(items);
    }

    getById(id: string): Observable<BacklogItem> {
        const items = this.readAll();
        const item = items.find(i => i.id === id);
        if (!item) return throwError(() => ({ error: { error: 'Item not found.' } }));
        const inPlanIds = this.getInPlanIds();
        return of({ ...item, isInActivePlan: inPlanIds.has(item.id) });
    }

    create(item: { title: string; description?: string; category: BacklogCategory; estimatedHours: number }): Observable<BacklogItem> {
        if (!item.title || !item.title.trim()) {
            return throwError(() => ({ error: { error: 'Title cannot be empty.' } }));
        }
        if (!item.estimatedHours || item.estimatedHours <= 0) {
            return throwError(() => ({ error: { error: 'Estimated hours must be > 0.' } }));
        }
        const items = this.readAll();
        const newItem: BacklogItem = {
            id: this.uuid(),
            title: item.title.trim(),
            description: item.description || '',
            category: item.category,
            estimatedHours: item.estimatedHours,
            isArchived: false,
            isInActivePlan: false,
            createdAt: new Date().toISOString()
        };
        items.push(newItem);
        this.writeAll(items);
        return of(newItem);
    }

    update(id: string, item: { title: string; description?: string; category: BacklogCategory; estimatedHours: number }): Observable<BacklogItem> {
        const items = this.readAll();
        const idx = items.findIndex(i => i.id === id);
        if (idx === -1) return throwError(() => ({ error: { error: 'Item not found.' } }));
        items[idx] = {
            ...items[idx],
            title: item.title,
            description: item.description || '',
            category: item.category,
            estimatedHours: item.estimatedHours
        };
        this.writeAll(items);
        return of(items[idx]);
    }

    archive(id: string): Observable<void> {
        const items = this.readAll();
        const idx = items.findIndex(i => i.id === id);
        if (idx === -1) return throwError(() => ({ error: { error: 'Item not found.' } }));
        items[idx] = { ...items[idx], isArchived: true };
        this.writeAll(items);
        return of(void 0);
    }

    delete(id: string): Observable<void> {
        let items = this.readAll();
        items = items.filter(i => i.id !== id);
        this.writeAll(items);
        return of(void 0);
    }
}

File: backend/src/WeeklyPlanTracker.Core/Services/BacklogService.cs
using WeeklyPlanTracker.Core.Entities;
using WeeklyPlanTracker.Core.Enums;
using WeeklyPlanTracker.Core.Interfaces;

namespace WeeklyPlanTracker.Core.Services;

/// <summary>Business logic for managing backlog items.</summary>
public class BacklogService
{
    private readonly IUnitOfWork _uow;

    public BacklogService(IUnitOfWork uow) => _uow = uow;

    /// <summary>Returns backlog items with optional filters.</summary>
    public Task<IEnumerable<BacklogItem>> GetAllAsync(
        bool includeArchived = false,
        BacklogCategory? category = null,
        string? search = null) =>
        _uow.BacklogItems.GetAllAsync(includeArchived, category, search);

    /// <summary>Returns a single backlog item by ID.</summary>
    public Task<BacklogItem?> GetByIdAsync(Guid id) =>
        _uow.BacklogItems.GetByIdAsync(id);

    /// <summary>Creates a new backlog item.</summary>
    public async Task<BacklogItem> CreateAsync(string title, string? description, BacklogCategory category, int estimatedHours)
    {
        if (string.IsNullOrWhiteSpace(title))
            throw new InvalidOperationException("Title cannot be empty.");

        if (estimatedHours <= 0)
            throw new InvalidOperationException("Estimated hours must be greater than 0.");

        var item = new BacklogItem
        {
            Title = title.Trim(),
            Description = description?.Trim(),
            Category = category,
            EstimatedHours = estimatedHours
        };

        _uow.BacklogItems.Add(item);
        await _uow.SaveChangesAsync();
        return item;
    }

    /// <summary>Updates an existing backlog item's properties.</summary>
    public async Task<BacklogItem> UpdateAsync(Guid id, string title, string? description, BacklogCategory category, int estimatedHours)
    {
        if (string.IsNullOrWhiteSpace(title))
            throw new InvalidOperationException("Title cannot be empty.");

        if (estimatedHours <= 0)
            throw new InvalidOperationException("Estimated hours must be greater than 0.");

        var item = await _uow.BacklogItems.GetByIdAsync(id)
            ?? throw new InvalidOperationException("Backlog item not found.");

        item.Title = title.Trim();
        item.Description = description?.Trim();
        item.Category = category;
        item.EstimatedHours = estimatedHours;

        _uow.BacklogItems.Update(item);
        await _uow.SaveChangesAsync();
        return item;
    }

    /// <summary>Archives a backlog item so it no longer appears in the planning picker.</summary>
    public async Task ArchiveAsync(Guid id)
    {
        var item = await _uow.BacklogItems.GetByIdAsync(id)
            ?? throw new InvalidOperationException("Backlog item not found.");

        item.IsArchived = true;
        _uow.BacklogItems.Update(item);
        await _uow.SaveChangesAsync();
    }

    /// <summary>Permanently deletes a backlog item.</summary>
    public async Task DeleteAsync(Guid id)
    {
        var item = await _uow.BacklogItems.GetByIdAsync(id)
            ?? throw new InvalidOperationException("Backlog item not found.");

        _uow.BacklogItems.Remove(item);
        await _uow.SaveChangesAsync();
    }

    /// <summary>Returns IDs of all backlog items currently assigned in the given weekly plan.</summary>
    public Task<IEnumerable<Guid>> GetAssignedItemIdsAsync(Guid weeklyPlanId) =>
        _uow.BacklogItems.GetAssignedItemIdsForPlanAsync(weeklyPlanId);
}
